pkg/sitewise/json_mapping.go

```diff
@@ -0,0 +1,182 @@
+package sitewise
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/grafana/iot-sitewise-datasource/pkg/models"
+	"github.com/grafana/iot-sitewise-datasource/pkg/sitewise/resource"
+)
+
+func BuildPropertyNameMap(ctx context.Context, resources resource.ResourceProvider, assetID string) map[string]string {
+	propertyNameMap := map[string]string{}
+	if assetID == "" {
+		backend.Logger.Warn("BuildPropertyNameMap: empty assetId")
+		return propertyNameMap
+	}
+
+	modelResp, err := resources.AssetModel(ctx)
+
+	if err != nil {
+		backend.Logger.Warn("DescribeAssetModel failed", "err", err)
+		return propertyNameMap
+	}
+
+	for _, prop := range modelResp.AssetModelProperties {
+		if prop.Id != nil && prop.Name != nil {
+			propertyNameMap[*prop.Id] = *prop.Name
+		}
+	}
+
+	if len(propertyNameMap) == 0 {
+		backend.Logger.Warn("BuildPropertyNameMap: No properties found for asset", "assetID", assetID)
+	}
+
+	return propertyNameMap
+}
+
+func isRequireJSONParsing(query models.BaseQuery) bool {
+	switch query.QueryType {
+	case models.QueryTypePropertyValueHistory,
+		models.QueryTypePropertyAggregate,
+		models.QueryTypePropertyValue:
+		return true
+	default:
+		return false
+	}
+}
+
+func ParseJSONFields(
+	ctx context.Context,
+	frames data.Frames,
+	resources resource.ResourceProvider,
+	assetID string,
+) data.Frames {
+	backend.Logger.Info("ParseJSONFields: starting JSON parsing", "assetID", assetID)
+
+	// Build property name map once
+	propertyNameMap := BuildPropertyNameMap(ctx, resources, assetID)
+	newFrames := data.Frames{}
+
+	for _, frame := range frames {
+		newFields := make([]*data.Field, 0, len(frame.Fields))
+		jsonParsed := false
+
+		for _, field := range frame.Fields {
+			newFields = append(newFields, field)
+
+			if field.Type() != data.FieldTypeString || field.Len() == 0 {
+				continue
+			}
+
+			rowCount := field.Len()
+			jsonFields := map[string]*data.Field{}
+
+			for r := 0; r < rowCount; r++ {
+				rawStr, _ := field.At(r).(string)
+				if rawStr == "" {
+					continue
+				}
+
+				var obj map[string]interface{}
+				if err := json.Unmarshal([]byte(rawStr), &obj); err != nil {
+					backend.Logger.Warn("ParseJSONFields: corrupted JSON, skipping row", "err", err, "frame", frame.Name, "row", r)
+					continue
+				}
+				jsonParsed = true
+				// Required fields check
+				for _, req := range []string{"timestamp", "prediction", "prediction_reason"} {
+					if _, ok := obj[req]; !ok {
+						backend.Logger.Warn("ParseJSONFields: missing required field", "field", req, "frame", frame.Name, "row", r)
+					}
+				}
+
+				for key, val := range obj {
+					if key == "diagnostics" || key == "timestamp" {
```

Comment:
shouldn't we also skip anomaly score if we handle it in the next block? Also are we intentionally skipping timestamp?